etna/models/nn/deepar_new/deepar.py

```diff
@@ -0,0 +1,372 @@
+        optimizer_params: Optional[dict],
+    ) -> None:
+        """Init DeepAR.
+
+        Parameters
+        ----------
+        input_size:
+            size of the input feature space: target plus extra features
+        num_layers:
+            number of layers
+        hidden_size:
+            size of the hidden state
+        lr:
+            learning rate
+        scale:
+            if True, scale target values in batch before training by :math:`1 + mean`, where :math:`mean` is mean of target values in the encoder part of batch
+        n_samples:
+            if 1, return theoretical mean of distribution as a predicted value. if greater than 1, return mean of `n_samples` generated from Monte Carlo sampling
+        loss:
+            loss function
+        optimizer_params:
+            parameters for optimizer for Adam optimizer (api reference :py:class:`torch.optim.Adam`)
+        """
+        super().__init__()
+        self.save_hyperparameters()
+        self.input_size = input_size
+        self.num_layers = num_layers
+        self.hidden_size = hidden_size
+        self.lr = lr
+        self.scale = scale
+        self.n_samples = n_samples
+        self.loss = loss
+        self.optimizer_params = {} if optimizer_params is None else optimizer_params
+        self.rnn = nn.LSTM(
+            num_layers=self.num_layers, hidden_size=self.hidden_size, input_size=self.input_size, batch_first=True
+        )
+        self.projection = nn.ModuleDict(
+            {
+                "loc": nn.Linear(in_features=self.hidden_size, out_features=1),
+                "scale": nn.Linear(in_features=self.hidden_size, out_features=1),
+            }
+        )
+
+    def forward(self, x: DeepARBatchNew, *args, **kwargs):  # type: ignore
+        """Forward pass.
+
+        Parameters
+        ----------
+        x:
+            batch of data
+
+        Returns
+        -------
+        :
+            forecast with shape (batch_size, decoder_length, 1)
+        """
+        encoder_real = x["encoder_real"].float()  # (batch_size, encoder_length-1, input_size)
+        decoder_real = x["decoder_real"].float()  # (batch_size, decoder_length, input_size)
+        decoder_target = x["decoder_target"].float()  # (batch_size, decoder_length, 1)
+        decoder_length = decoder_real.shape[1]
+        weights = x["weight"]
+        forecasts = torch.zeros((decoder_target.shape[0], decoder_target.shape[1], self.n_samples))
+
+        for j in range(self.n_samples):
+            _, (h_n, c_n) = self.rnn(encoder_real)
+            for i in range(decoder_length):
+                output, (h_n, c_n) = self.rnn(decoder_real[:, i, None], (h_n, c_n))  # (batch_size, 1, hidden_size)
+                loc, scale = self.get_distribution_params(output[:, -1])
+                forecast_point = self.loss.sample(
+                    loc=loc, scale=scale, weights=weights, theoretical_mean=True if self.n_samples == 1 else False
```

Comment:
Better do define `_theoretical_mean` attribute in the constructor as it is some kind of property of the network